doc/dev/design/NAT-address-discovery.rst

```diff
@@ -0,0 +1,163 @@
+One example of this approach is the `STUN protocol <https://en.wikipedia.org/wiki/STUN>`_.
+
+One complication is that the local AS of the sender may be split into different subnets and that border routers
+are not all in the same subnet.
+Another complication is that some NATs may change their port mapping port if a client endhost connects to a new
+remote IP or port or uses a different local port.
+
+Therefore, it is desirable for the detector to be not only in the same AS and same subnet as the BR, but ideally at the
+same IP address, listening on the same port.
+Similarly, it is desirable for the client to use the same local port when connecting to the detector and for actual data
+packets to the border router.
+
+Separately, we need to specify a way how the client endhost can discover the "detector", i.e. its IP address and port.
+In case the detector shares and IP and possibly even port (or uses a fixed port) with the border router, this
+discovery is straight forward.
+If the port is flexible or the detector runs on a different IP then we need to find a different solution, probably
+as an extension to the bootstrapping service.
+
+Proposal
+========
+[A precise statement of the proposed change.]
+
+Proposed change: Extend the border router to detect NATed addresses/ports and report them back to a client endhost.
+
+Ideally the solution would listen on the same port that would also be used for normal traffic forwarding.
+Alternatively, it may be feasible (depending on how sensitive the NAT is) to use a different (fixed) port to
+accept requests for reporting addresses.
+
+The implementation on the protocol level could be done in several ways:
+
+1.  STUN: The BR needs to detect if an incoming packet is a STUN packet, and if it is, respond to the STUN request.
+
+    -  Advantages: STUN is a well known and mature protocol. There are STUN libraries available in many programming languages.
+       It should be easy to implement this.
+
+2.  Extend SCMP with a new INFO message
+
+    -  Advantages: One less dependency on an external library and protocol
+    -  Disadvantages: More standardization effort? How do we solve authentication?
+
+3.  Extend SCMP with a new ERROR message: "invalid source address for first hop pkt", similar to error 33.
+    The router can verify that for first hop packets, the IP src address (and L4 port if applicable) matches the SCION
+    src address (and L4 port).
+    If not, it returns an error. The actual source address would need to be attached somewhere, unless we decide
+    to change the payload so it contains the IP header of the offending packet (and the IP header should contain the
+    NATed IP/port).
+
+    -  Advantages: One less dependency on an external library and protocol
+    -  Disadvantages: Conceptually a bit of a hack. The BR would need to check every outbound packet as part of the fast
+       path. More standardization effort? How do we solve authentication?
+
```

Comment:
To avoid potential misunderstandings make it here explicit, that this proposal only addresses the problem of SCION clients behind NATs. It is not the goal to also support NAT traversal techniques such as NAT hole punching SCION-based servers and peer-to-peer scenarios.